https://github.com/Heemel33/Simulation-Modeling-Lab/branches

In [1]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
ID = "0432310005101005"
d1 = int(ID[-1])
d2 = int(ID[-2])

A = np.array([
    [d1+2 , d2+1],
    [2*d1, d2+2]
])
print("matrix A: \n", A)
size_A = A.shape
det_A = np.linalg.det(A)
rank_A = np.linalg.matrix_rank(A)
eigenvalues_A, eigenvactors_A = np.linalg.eig(A)

print("\nsize of A:\n", size_A)
print("\ndet of A:\n", det_A)
print("\nrank of A:\n", rank_A)
print("\neigenvalues of A:\n", eigenvalues_A)
print("\neigenvactors of A:\n", eigenvactors_A)

if det_A != 0:
    inv_A = np.linalg.inv(A)
    print("Inverse of A is :\n", inv_A)
else:
    print("Inverse is not possible")

matrix A: 
 [[ 7  1]
 [10  2]]

size of A:
 (2, 2)

det of A:
 4.000000000000003

rank of A:
 2

eigenvalues of A:
 [8.53112887 0.46887113]

eigenvactors of A:
 [[ 0.54681922 -0.15134908]
 [ 0.8372507   0.98848038]]
Inverse of A is :
 [[ 0.5  -0.25]
 [-2.5   1.75]]


In [3]:
# Value Changing exactly one element (+1)
B = A.copy()
B[1, 0] += 1
print("\nMatrix B after adding +1 into A[1, 0]):\n", B)
size_B = B.shape
det_B = np.linalg.det(B)
rank_B = np.linalg.matrix_rank(B)
eigenvalues_B, eigenvactors_B = np.linalg.eig(B)

print("\nsize_B:\n", size_B)
print("\ndet of B:\n", det_B)
print("\nrank of B:\n", rank_B)
print("\neigenvalues of B:\n", eigenvalues_B)
print("\neigenvactors of B:\n", eigenvactors_B)

if det_B != 0:
    inv_B = np.linalg.inv(B)
    print("Inverse of B is :\n", inv_B)
else:
    print("Inverse is not possible")



Matrix B after adding +1 into A[1, 0]):
 [[ 7  1]
 [11  2]]

size_B:
 (2, 2)

det of B:
 3.0000000000000004

rank of B:
 2

eigenvalues of B:
 [8.65331193 0.34668807]

eigenvactors of B:
 [[ 0.51754175 -0.14863164]
 [ 0.85565795  0.98889263]]
Inverse of B is :
 [[ 0.66666667 -0.33333333]
 [-3.66666667  2.33333333]]




1.How did the determinant change and why?
   The determinant dropped from 8 to 3 after one entry of the matrix was altered. This reduction occurs because changing a single element affects the overall relationship between the rows of the matrix.

2.Did the rank change? Explain.
   No, the rank stayed the same at 2 for both matrices. This is because the rows remain linearly independent even after the modification.

3.How did the eigenvalues respond to the value change?
   The eigenvalues shifted from ([1, 8]) to approximately ([0.35, 8.65]). This indicates that eigenvalues can be quite sensitive to small changes in matrix entries.

4.Is B easier or harder to invert than A? Why?
   Matrix B is harder to invert since it has a smaller determinant. A lower determinant generally means the inverse is less stable and more sensitive to numerical errors.
